### DISCUSSION TITLE
Improve Web Search with by preprocessing search query

### DISCUSSION THREAD

**Discussion by Expro** [open-webui/open-webui#2822] [General]
**Is your feature request related to a problem? Please describe.**

Currently, Web Search functionality simply forwards prompt that user put into chat toward search engine. This is not ideal, as chat encourages conversation-type expressions, while search engines prefers keyword-like expressions.

**Describe the solution you'd like**

It would be great if user prompt could be preprocessed before being being send to search engine. It could be as simple as allowing user to specify preprocessing system prompt for model that would be used for web search prepocessing.

Real world example:

Today I was testing Web Search functionality by asking LLama3 3 following prompt:

"What Azure services can be deployed on Azure HCI?"

This of course resulted in my SearXNG instance being queried with query "What Azure services can be deployed on Azure HCI?". Results were not bad, but they could be better.

I ran a simple experiment: in separate chat I've prompted LLama3 with following expression:

"Turn following text into search-friendly expression and answer with expression only, without quotation marks:

What Azure services can be deployed on Azure HCI?"

Llama 3 answered with: "azure hci supported services". Then I've used that answer as prompt with Web Search and received better answer.

Such solution would allow for natural, conversation-like prompts with search-friendly queries being used for web searches.

**Comment by cheahjs** [comment]
There's currently a hardcoded threshold at 100 characters before it queries a model for a search phrase, maybe we could make this threshold configurable?

**Reply by Expro** [reply]
Good idea, with "0" also meaning "always query model".

Configurable system prompt for search phrase queries would be also nice addition.

**Reply by alkeryn** [reply]
does it uses the conversation context ?
because if it does that'd be nice especially to properly reply to subsequent question or questions that depends on user information.
also if the model is always queryied we may be able to skip searching when it's not necessary.

**Reply by Expro** [reply]
One additional comment: after giving it some thought, system prompt for search phrase queries seems to be necessity for some scenarios.

For example, if I create model with system prompt telling it to impersonate someone (for example Albert Einstein) and I try to chat with it as with person, Web Search breaks down. If I ask Albert Einstein "Tell me something about yourself", it will forward "Tell me something about yourself" to search engine, returning data not associated with Albert Einstein. With system prompt for search queries I could ask model to reformat user prompt to include info that question is about Albert Einstein, making Web Search actually useful.

**Reply by alkeryn** [reply]
yes there will definitely needs to be some system prompting but the gist of it is that it should be able to use its context to create the search query.

**Reply by rcdailey** [reply]
I'm also personally interested in this. I think it's needed to generate search queries using previous context. Here's a test I did:

Prompt: Summarize the current happenings of the 2024 presidential election
Response: Provides information about Biden vs Trump (which is outdated news since as of right now Biden has dropped out, so the news should instead be mentioning Harris)

Prompt: That information is out of date
Response: Could you tell me what information is out of date?

Basically, the "That information is out of date" was sent to the search engine, instead of Open WebUI using past conversation to build a search query intelligently to query for more recent information related to the election (my first prompt).
